### PR TITLE
"Refacto payment Service"

### PR DESCRIPTION
**Pull request by charlespernet**

### CODE REVIEW

File: app/controllers/credit_cards_controller.rb [hunk 1]
```diff
@@ -2,39 +2,43 @@ class CreditCardsController < ApplicationController
   # Try a better approach than skiping once done
   skip_after_action :verify_policy_scoped
   skip_after_action :verify_authorized
-  before_action :set_payment_service
 
   def index
-    @stripe_sources = @payment_service.user_sources
+    @stripe_sources = current_user.credit_cards
```
Comment (fcatuhe) [inline app/controllers/credit_cards_controller.rb]: be more consistent by renaming @stripe_sources to @credit_cards?

Comment (fcatuhe) [reply app/controllers/credit_cards_controller.rb]: after all, we're in a CreditCardsController, seems obvious to have @credit_cards in #index action.

File: app/controllers/payments_controller.rb [hunk 2]
```diff
@@ -2,55 +2,55 @@ class PaymentsController < ApplicationController
...
-      else
-        redirect_to cookoons_path, flash: { payment_succeed: true }
-      end
+    payment = Reservation::Payment.new(@reservation, payment_params)
+    if payment.proceed
+      @reservation.notify_users_after_payment
+      handle_redirection
     else
-      flash.now.alert = payment_service.displayable_errors
+      @user_cards = current_user.credit_cards
```
Comment (fcatuhe) [inline app/controllers/payments_controller.rb]: @credit_cards here also? Does it transport more meaning than @user_cards?

File: app/controllers/services/payments_controller.rb [hunk 3]
```diff
@@ -4,37 +4,34 @@ class Services::PaymentsController < ApplicationController
...
-      },
-      discount: params[:payment][:use_discount]
-    )
-    @user_cards = payment_service.user_sources.try(:data)
-    if payment_service.pay_service
+    payment = Service::Payment.new(@service, service_params)
+    if payment.proceed
       redirect_to cookoons_path, flash: { service_payment_succeed: true }
     else
+      @user_cards = current_user.crredit_cardds
```
Comment (fcatuhe) [inline app/controllers/services/payments_controller.rb]: same here, @credit_cards?

Comment (fcatuhe) [reply app/controllers/services/payments_controller.rb]: Typo on crredit_cardds?

File: app/models/concerns/stripe/chargeable.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      return false unless charge
+      charge.capture
+    end
+
+    def charge
+      @charge ||= Stripe::Charge.retrieve(payee.stripe_charge_id)
+    end
+
+    def persist_charge
+      payee.update(stripe_charge_id: stripe_charge&.id)
```
Comment (fcatuhe) [inline app/models/concerns/stripe/chargeable.rb]: payee replaced chargeable? Maybe chargee?

File: app/models/concerns/stripe/chargeable.rb [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+      @stripe_charge = Stripe::Charge.create(charge_options)
+    rescue Stripe::CardError, Stripe::InvalidRequestError => e
+      Rails.logger.error('Failed to create Stripe Charge')
+      Rails.logger.error(e.message)
+      errors << e.message
+      # could instantiate a Payment::Unprocessable
+      false
+    end
+
+    def charge_options
```
Comment (fcatuhe) [inline app/models/concerns/stripe/chargeable.rb]: naming: charge_options or charge_attributes? I would go for attributes, e.g. amount is not optional

File: app/models/payment.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  delegate :user, :payment_amount_cents, to: :payee
+
+  def initialize(payee, options = {})
+    @payee = payee
+    @options = options
+    @errors = []
+  end
+
+  def proceed
+    # #discount_asked? and #charge_needed? sit within #discountable
```
Comment (fcatuhe) [inline app/models/payment.rb]: should we move discount_asked? and charge_needed? in this model? Logic would be: when a model wants to be discountable, he must provide these 2 public methods?

File: app/models/concerns/discountable.rb [hunk 7]
```diff
@@ -0,0 +1,61 @@
+module Discountable
+  def computed_discount_amount_cents
```
Comment (fcatuhe) [inline app/models/concerns/discountable.rb]: maybe we should have a charge_amount_cents in Payment, that would either be = payment_amount_cents if Payment is not discountable, or defined differently if discountable (charge_amount_cents = discountable_charge_amount_cents in Payment for example)?
discountable_charge_amount_cents would be the actual computed_charge_amount_cents

File: app/models/concerns/stripe/chargeable.rb (hunk 4 above)
Comment (charlespernet) [reply app/models/concerns/stripe/chargeable.rb]: chargeable

File: app/models/concerns/stripe/chargeable.rb (hunk 5 above)
Comment (charlespernet) [reply app/models/concerns/stripe/chargeable.rb]: charge_attributes

Comment (charlespernet) [reply app/models/concerns/stripe/chargeable.rb]: charge_attributes

File: app/models/payment.rb (hunk 6 above)
Comment (charlespernet) [reply app/models/payment.rb]: Remonter en methode publique discount_asked? dans ::Payment 

charge_needed? doit etre remonté dans ::Payment

def charge_needed?
    charge_amount_cents.positive?
  end

Comment (charlespernet) [reply app/models/payment.rb]: charge_amount_cents doit etre implémenter dans tous les Payment

File: app/models/concerns/stripe/customerable.rb [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+    private
+
+    def update_user(customer)
+      user.update(stripe_customer_id: customer.id) if customer
+    end
+
+    def retrieve_customer
+      Stripe::Customer.retrieve(stripe_customer_id)
+    rescue Stripe::InvalidRequestError => e
```
[the remark this answers is not in the document]
Comment (charlespernet) [reply app/models/concerns/stripe/customerable.rb]: Je vais essayer de regarder dans cette direction. Je pense que les actions de paiement devraient globalement raise une Exception et le controller devrait la rescue ! Je pense que ca serait plus logique. Qu'en penses tu ?

**Pull request merged**